registry/storage/walk.go

```diff
@@ -0,0 +1,59 @@
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	storageDriver "github.com/docker/distribution/registry/storage/driver"
+)
+
+// basename returns the filename and path components of a path
+func basename(path string) (string, []string) {
+	pathComponents := strings.Split(path, "/")
+	baseName := pathComponents[len(pathComponents)-1]
+	rest := pathComponents[:len(pathComponents)-1]
+	return baseName, rest
+}
+
+// SkipDir is used as a return value from onFileFunc to indicate that
+// the directory named in the call is to be skipped. It is not returned
+// as an error by any function.
+var ErrSkipDir = errors.New("skip this directory")
+
+// onFileFunc is called once per file by Walk
+// If the returned error is ErrSkipDir and fileInfo refers
+// to a directory, the directory will not be entered and Walk
+// will continue the traversal.  Otherwise Walk will return
+type onFileFunc func(fileInfo storageDriver.FileInfo) error
+
+// Walk traverses a filesystem defined within driver, starting
+// from the given path, calling f on each file
+func Walk(driver storageDriver.StorageDriver, from string, f onFileFunc) error {
```

Comment:
Looking at this closely, it actually belongs in the driver package. This provides fs walks over driver implementations and doesn't need to have knowledge of storage fundamentals.